src/Compilers/CSharp/Test/Emit/CodeGen/UnsafeTests.cs

```diff
@@ -3985,6 +3985,1041 @@ .maxstack  2
+}
+
+";
+
+            var compVerifier = CreateCompilationWithMscorlib46(text, options: TestOptions.UnsafeReleaseExe);
+
+            compVerifier.VerifyDiagnostics(
+                // (8,25): error CS9365: The given expression cannot be used in a fixed statement
+                //         fixed (int* p = f)
+                Diagnostic(ErrorCode.ERR_ExprCannotBeFixed, "f").WithLocation(8, 25)
+                );
+        }
+
+        [Fact]
+        public void SimpleCaseOfCustomFixedGeneric()
+        {
+            var text = @"
+unsafe class C
+{
+    public static void Main()
+    {
+        Test(42);
+        Test((object)null);
+    }
+
+    public static void Test<T>(T arg)
+    {
+        fixed (int* p = arg)
+        {
+            System.Console.Write(p == null? 0: p[1]);
+        }
+    }
+}
+
+static class FixAllExt
+{
+    public static ref int DangerousGetPinnableReference<T>(this T dummy)
+    {
+        return ref (new int[]{1,2,3})[0];
+    }
+}
+";
+
+            var compVerifier = CompileAndVerify(text, additionalRefs: new[] { ExtensionAssemblyRef },  options: TestOptions.UnsafeReleaseExe, expectedOutput: @"20", verify: Verification.Fails);
+
+            compVerifier.VerifyIL("C.Test<T>(T)", @"
+{
+  // Code size       49 (0x31)
+  .maxstack  2
+  .locals init (int* V_0, //p
+                pinned int& V_1)
+  IL_0000:  ldarg.0
+  IL_0001:  box        ""T""
+  IL_0006:  brtrue.s   IL_000c
+  IL_0008:  ldc.i4.0
+  IL_0009:  conv.u
+  IL_000a:  br.s       IL_001b
+  IL_000c:  ldarga.s   V_0
+  IL_000e:  ldobj      ""T""
+  IL_0013:  call       ""ref int FixAllExt.DangerousGetPinnableReference<T>(T)""
+  IL_0018:  stloc.1
+  IL_0019:  ldloc.1
+  IL_001a:  conv.u
+  IL_001b:  stloc.0
+  IL_001c:  ldloc.0
+  IL_001d:  ldc.i4.0
+  IL_001e:  conv.u
+  IL_001f:  beq.s      IL_0027
+  IL_0021:  ldloc.0
+  IL_0022:  ldc.i4.4
+  IL_0023:  add
+  IL_0024:  ldind.i4
+  IL_0025:  br.s       IL_0028
+  IL_0027:  ldc.i4.0
+  IL_0028:  call       ""void System.Console.Write(int)""
+  IL_002d:  ldc.i4.0
+  IL_002e:  conv.u
+  IL_002f:  stloc.1
+  IL_0030:  ret
+}
+");
+        }
+
+        [Fact]
+        public void CustomFixedStructSideeffects()
+        {
+            var text = @"
+    unsafe class C
+    {
+        public static void Main()
+        {
+            var b = new FixableStruct();
+            Test(ref b);
+            System.Console.WriteLine(b.x);
+        }
+
+        public static void Test(ref FixableStruct arg)
+        {
+            fixed (int* p = arg)
+            {
+                System.Console.Write(p[1]);
+            }
+        }
+    }
+
+    struct FixableStruct
+    {
+        public int x;
+
+        public ref int DangerousGetPinnableReference()
+        {
+            x = 456;
+            return ref (new int[] { 4, 5, 6 })[0];
+        }
+    }
+";
+
+            var compVerifier = CompileAndVerify(text, options: TestOptions.UnsafeReleaseExe, verify: Verification.Fails, expectedOutput: @"5456");
```

Comment:
>verify: Verification.Fails [](start = 93, length = 26)

Would this be verifiable by some updated verification rules?

Reply:
The ref-returning method will be, but any use of `int *` will not

---
In reply to: [167672486](https://github.com/dotnet/roslyn/pull/24469#discussion_r167672486) [](ancestors = 167672486)